lib/registries/addon/services/search.ts

```diff
@@ -0,0 +1,184 @@
+/* tslint:disable max-classes-per-file */
+import Service from '@ember/service';
+import { Map, OrderedSet, ValueObject } from 'immutable';
+import $ from 'jquery';
+
+// Used later on for slightly stricter typing of immutable Maps
+// Records don't conform to the same interface as Map, so it may not be used as
+// a key in other Maps
+interface TypedMap<T> extends Map<keyof T, T[keyof T]> {
+    get<K extends keyof T>(key: K): T[K] | undefined;
+    set<K extends keyof T>(key: K, value: T[K]): this;
+}
+
+export class MapProxy<T> implements ValueObject {
+    protected readonly internal: TypedMap<T>;
+
+    get hashable(): TypedMap<T> {
+        return this.internal;
+    }
+
+    constructor(values: T) {
+        this.internal = Map(Object.entries(values)) as TypedMap<T>;
+    }
+
+    equals(other: any): boolean {
+        return this.internal.equals(other instanceof MapProxy ? other.hashable : other);
+    }
+
+    hashCode(): number {
+        return this.internal.hashCode();
+    }
+}
+
+// Doesn't matter for types but useful for keeping all interfaces together.
+/* eslint-disable no-use-before-define */
+
+interface Filter {
+   key: string;
+   type: symbol;
+   display: string;
+   value: string | number | boolean;
+}
+
+interface Order {
+    display: string;
+    key?: string;
+    ascending: boolean;
+}
+
+interface Options {
+    filters: OrderedSet<SearchFilter>;
+    modifiers: OrderedSet<SearchModifier>;
+    order: SearchOrder;
+    page: number;
+    query?: string;
+    size: number;
+}
+
+export interface SearchModifier extends ValueObject {
+    apply(query: any): any;
+}
+
+/* eslint-enable no-use-before-define */
+
+export class SearchOrder extends MapProxy<Order> implements Order {
+    get display() {
+        return this.internal.get('display')!;
+    }
+
+    get key() {
+        return this.internal.get('key');
+    }
+
+    get ascending() {
+        return this.internal.get('ascending')!;
+    }
+}
+
+export abstract class SearchFilter extends MapProxy<Filter> implements Filter, SearchModifier {
+    get key() {
```

Comment:
I think I like these immutable.js patterns! They're clean and give a comfy feeling of safety.

Super minor: I kinda wish they weren't so verbose. Is there a way we could replace
```ts
get key() {
    return this.internal.get('key')!;
}
```
with something small? It'd be nice if a decorator could help. Property decorator wouldn't work, but could maybe do an accessor decorator?
```ts
@proxyGet
get key() {
    // return default value? or raise?
}

// could be one-lined
@proxyGet get foo { ... }
```
...but that's not really better. A class decorator could maybe do:
```ts
@proxyAttrs('key', 'type', 'display', 'value')
class SearchFilter extends MapProxy<Filter> implements Filter, SearchModifier {
    abstract apply(query: any): any;
}
```
I *think* that's possible, but could be weird.

Reply:
I was under the impression that decorators couldn't mutate the actual type of the class?

From https://www.typescriptlang.org/docs/handbook/decorators.html
> Decorators use the form @expression, where expression must evaluate to a function that will be called at runtime with information about the decorated declaration

Reply:
Property decorators can't do anything to the property, just observe that it happened and maybe do something to the class prototype.

Accessor and method decorators get a descriptor, which they can modify, so you can wrap a method in another function, or replace a get/set with something else.

Class decorators are super powerful; they can [replace the class wholesale](https://www.typescriptlang.org/docs/handbook/decorators.html#class-decorators):
> If the class decorator returns a value, it will replace the class declaration with the provided constructor function.

We use this [in `@requireAuth`](https://github.com/CenterForOpenScience/ember-osf-web/blob/develop/app/decorators/require-auth.ts#L27) to insert a `beforeModel` hook. Though with great power comes great responsibility, and it might be hard to tell when we start abusing it..